apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/server/AbstractRandomClaimGenerator.java

```diff
@@ -84,73 +183,254 @@ protected String randomDate() {
    *
    * @param values all possible values
-   * @param <T> the enum type
+   * @param <TEnum> the enum type
    * @return one of the values selected at random
    */
-  protected <T> T randomEnum(List<T> values) {
-    return values.get(random.nextInt(values.size()));
+  protected <TEnum> TEnum randomEnum(List<TEnum> values) {
+    RandomValueContext ctx = createContext();
+    return values.get(ctx.randomInteger(values.size()));
+  }
+
+  /**
+   * Creates a {@link RandomValueContext} using the current {@link #seed}, {@link #sequence}, and
+   * {@link #path} attributes that can be used to generate random values.
+   *
+   * @return The created {@link RandomValueContext}.
+   */
+  private RandomValueContext createContext() {
+    return createContext(null);
+  }
+
+  /**
+   * Creates a {@link RandomValueContext} using the current {@link #seed}, {@link #sequence}, and
+   * {@link #path} attributes that can be used to generate random values.
+   *
+   * @param prefix An optional (nullable) prefix that can be used for accessory random values, such
+   *     as determining if an optional() value should be added.
+   * @return The created {@link RandomValueContext}.
+   */
+  private RandomValueContext createContext(String prefix) {
+    String prefixString = prefix != null && !prefix.isBlank() ? prefix + "." : "";
+    String propertyPath =
+        prefixString + path.stream().filter(Objects::nonNull).collect(Collectors.joining("."));
+    return new RandomValueContext(seed + sequence + propertyPath.hashCode());
+  }
+
+  /**
+   * Helper method to add a new layer to the current {@link #path}, returning a {@link PathLayer}
+   * object that can be used to reset the path again to just before the layer was applied.
+   *
+   * @param propertyName The property name to add to the path.
+   * @return A {@link PathLayer} object that can be used to {@link PathLayer#remove()} the added
+   *     layer.
+   */
+  private PathLayer addLayer(String propertyName) {
+    if (propertyName != null && !propertyName.isBlank()) {
+      path.push(propertyName);
+      return path::pop;
+    }
+
+    return () -> {
+      // Don't pop any values if we didn't add any
+    };
   }
 
   /**
-   * Triggers the action 50% of the time (or always if optionalTrue has been set).
+   * Adds a new layer to the {@link #path} using the propertyName, which will affect the random
+   * values generated in the given action, then resets the {@link #path} to its prior state just
+   * before this method was invoked.
    *
-   * @param action action to trigger half the time
+   * @param propertyName The property name to add to the {@link #path}.
+   * @param action The {@link Runnable} action to execute, which should include calls methods within
+   *     this class to generate random values.
    */
-  protected void optional(Runnable action) {
-    if (optionalTrue || random.nextBoolean()) {
+  protected void always(String propertyName, Runnable action) {
```

Comment:
Is the always/optional naming scheme following a specific standard? IMO like something like addOptional and addRequired would be a little clearer on what the method is doing. (short punchy names are nice, but reading them in context in the next class it wasnt immediately clear to me what the method `always` was doing)

Reply:
They are wrappers for runnable logic execution, affecting if/when the logic is executed, while additionally setting a context to use for random value generation.  The logic itself doesn't need to be adding data to an object, for instance the `always()` method is used in the `createRandomClaim()` method merely to set the context for subsequent logic.  There may be a better naming system to use to indicate this, if someone has a better suggestion.

Reply:
Had some suggestions above, but not a blocking concern.